File: ios-app/Shared/TerseAuth.swift
import Foundation
import AuthenticationServices

class TerseAuth: ObservableObject {
    static let shared = TerseAuth()

    private let apiBase = "https://www.pruneai.com"
    private let defaults: UserDefaults
    private var webAuthSession: ASWebAuthenticationSession?
    private let contextProvider = WebAuthContextProvider()

    @Published var isSignedIn: Bool = false
    @Published var signInError: String?
    @Published var clerkUserId: String?
    @Published var email: String?
    @Published var firstName: String?
    @Published var imageUrl: String?

    // License
    @Published var tier: String = "free"
    @Published var status: String = "active"
    @Published var weeklyUsage: Int = 0
    @Published var optimizationsPerWeek: Int = 120
    @Published var maxSessions: Int = 1
    @Published var remaining: Int = 120

    var isUnlimited: Bool { optimizationsPerWeek < 0 }
    var canOptimize: Bool { isUnlimited || remaining > 0 }

    var usageText: String {
        if isUnlimited { return "Unlimited" }
        return "\(remaining)/\(optimizationsPerWeek) left this week"
    }

    var tierLabel: String {
        tier.prefix(1).uppercased() + tier.dropFirst()
    }

    init() {
        self.defaults = UserDefaults(suiteName: "group.com.pruneai.shared") ?? .standard
        loadAuth()
        loadLicense()
    }

    // MARK: - Sign in with Apple

    func signInWithApple(credential: ASAuthorizationAppleIDCredential) {
        guard let tokenData = credential.identityToken,
              let identityToken = String(data: tokenData, encoding: .utf8) else {
            print("[TerseAuth] No identity token from Apple")
            return
        }

        let appleEmail = credential.email
        let fullName = credential.fullName
        let firstName = fullName?.givenName
        let lastName = fullName?.familyName

        // Save Apple name locally (Apple only sends it on first sign-in)
        if let fn = firstName {
            defaults.set(fn, forKey: "appleFirstName")
        }
        if let ln = lastName {
            defaults.set(ln, forKey: "appleLastName")
        }

        let savedFirstName = firstName ?? defaults.string(forKey: "appleFirstName")
        let savedLastName = lastName ?? defaults.string(forKey: "appleLastName")

        guard let url = URL(string: "\(apiBase)/api/auth/apple") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        var body: [String: Any] = ["identityToken": identityToken]
        if let e = appleEmail { body["email"] = e }
        if let fn = savedFirstName { body["firstName"] = fn }
        if let ln = savedLastName { body["lastName"] = ln }

        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        print("[TerseAuth] Sending Apple token to server...")
        DispatchQueue.main.async { self.signInError = nil }

        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            guard let self = self else { return }

            if let error = error {
                print("[TerseAuth] Apple auth network error: \(error.localizedDescription)")
                DispatchQueue.main.async { self.signInError = "Network error. Please try again." }
                return
            }

            let httpStatus = (response as? HTTPURLResponse)?.statusCode ?? 0
            let rawStr = data.flatMap { String(data: $0, encoding: .utf8) } ?? "nil"
            print("[TerseAuth] Apple auth response (\(httpStatus)): \(rawStr)")

            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let status = json["status"] as? String,
                  status == "authenticated" else {
                print("[TerseAuth] Apple auth API failed or not authenticated")
                DispatchQueue.main.async { self.signInError = "Sign in failed. Please try again." }
                return
            }

            let userId = json["clerkUserId"] as? String ?? ""
            let email = json["email"] as? String ?? ""
            let imageUrl = json["imageUrl"] as? String
            let name = json["firstName"] as? String ?? savedFirstName

            print("[TerseAuth] Apple Sign In completed! User: \(userId)")

            DispatchQueue.main.async {
                self.signInError = nil
                self.saveAuth(clerkUserId: userId, email: email, imageUrl: imageUrl, firstName: name)
                self.verifyLicense()
            }
        }.resume()
    }

    // MARK: - Web Auth Flow (fallback)

    func startSignIn() {
        guard let url = URL(string: "\(apiBase)/api/auth/start") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"

        URLSession.shared.dataTask(with: request) { [weak self] data, _, error in
            guard let self = self,
                  let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let token = json["token"] as? String else { return }

            self.defaults.set(token, forKey: "authToken")
            self.defaults.set(false, forKey: "authCompleted")

            guard let authURL = URL(string: "\(self.apiBase)/auth-callback.html?token=\(token)&action=signin") else { return }

            DispatchQueue.main.async {
                let session = ASWebAuthenticationSession(
                    url: authURL,
                    callbackURLScheme: "terseauth"
                ) { _, _ in
                    // User dismissed or callback received — check if auth completed
                    self.checkPendingAuth()
                }
                session.presentationContextProvider = self.contextProvider
                session.prefersEphemeralWebBrowserSession = false
                session.start()
                self.webAuthSession = session
            }

            self.pollForAuth(token: token)
        }.resume()
    }

    /// Call this when the app returns to foreground to resume auth check
    func checkPendingAuth() {
        guard !isSignedIn else { return }
        guard let token = defaults.string(forKey: "authToken"), !token.isEmpty else { return }
        if defaults.bool(forKey: "authCompleted") { return }
        print("[TerseAuth] Resuming auth poll for token: \(token.prefix(8))...")
        pollForAuth(token: token)
    }

    private func pollForAuth(token: String) {
        DispatchQueue.global().asyncAfter(deadline: .now() + 1.0) { [weak self] in
            guard let self = self else { return }
            guard let url = URL(string: "\(self.apiBase)/api/auth/poll/\(token)") else { return }

            URLSession.shared.dataTask(with: url) { data, _, _ in
                guard let data = data,
                      let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let status = json["status"] as? String else {
                    // Retry
                    self.pollForAuth(token: token)
                    return
                }

                if status == "authenticated" {
                    let userId = json["clerkUserId"] as? String
                    let email = json["email"] as? String
                    let imageUrl = json["imageUrl"] as? String
                    let firstName = json["firstName"] as? String

                    self.defaults.set(true, forKey: "authCompleted")
                    print("[TerseAuth] Auth completed! User: \(userId ?? "nil")")

                    DispatchQueue.main.async {
                        self.webAuthSession?.cancel()
                        self.webAuthSession = nil
                        self.saveAuth(
                            clerkUserId: userId ?? "",
                            email: email ?? "",
                            imageUrl: imageUrl,
                            firstName: firstName
                        )
                        self.verifyLicense()
                    }
                } else if status == "pending" {
                    self.pollForAuth(token: token)
                }
            }.resume()
        }
    }

    // MARK: - Save/Load Auth

    private func saveAuth(clerkUserId: String, email: String, imageUrl: String?, firstName: String?) {
        self.clerkUserId = clerkUserId
        self.email = email
        self.imageUrl = imageUrl
        self.firstName = firstName
        self.isSignedIn = true

        defaults.set(true, forKey: "isSignedIn")
        defaults.set(clerkUserId, forKey: "clerkUserId")
        defaults.set(email, forKey: "email")
        defaults.set(imageUrl, forKey: "imageUrl")
        defaults.set(firstName, forKey: "firstName")
    }

    private func loadAuth() {
        isSignedIn = defaults.bool(forKey: "isSignedIn")
        clerkUserId = defaults.string(forKey: "clerkUserId")
        email = defaults.string(forKey: "email")
        imageUrl = defaults.string(forKey: "imageUrl")
        firstName = defaults.string(forKey: "firstName")
    }

    func signOut() {
        isSignedIn = false
        clerkUserId = nil
        email = nil
        imageUrl = nil
        firstName = nil

        defaults.removeObject(forKey: "isSignedIn")
        defaults.removeObject(forKey: "clerkUserId")
        defaults.removeObject(forKey: "email")
        defaults.removeObject(forKey: "imageUrl")
        defaults.removeObject(forKey: "firstName")

        // Reset license to free
        tier = "free"
        optimizationsPerWeek = 500
        remaining = 500
        weeklyUsage = 0
        saveLicense()
    }

    func deleteAccount() {
        guard let userId = clerkUserId, !userId.isEmpty else {
            signOut()
            return
        }

        // Call server to delete account
        guard let url = URL(string: "\(apiBase)/api/auth/delete") else {
            signOut()
            return
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: ["clerkUserId": userId])

        URLSession.shared.dataTask(with: request) { _, _, _ in
            print("[TerseAuth] Account deletion requested for \(userId)")
        }.resume()

        // Clear all local data
        signOut()
        defaults.removeObject(forKey: "stats_entries")
        defaults.removeObject(forKey: "totalTokensSaved")
        defaults.removeObject(forKey: "totalTokensOptimized")
        defaults.removeObject(forKey: "totalOptimizations")
        defaults.removeObject(forKey: "weeklyUsage")
        defaults.removeObject(forKey: "usageWeek")
        defaults.removeObject(forKey: "spentTokens")
        defaults.removeObject(forKey: "unlockedThemes")
        defaults.removeObject(forKey: "appleFirstName")
        defaults.removeObject(forKey: "appleLastName")
        defaults.synchronize()
    }

    // MARK: - License

    func verifyLicense() {
        guard let userId = clerkUserId, !userId.isEmpty else {
            print("[TerseAuth] No clerkUserId, skipping license verify")
            return
        }
        guard let url = URL(string: "\(apiBase)/api/license/\(userId)?platform=ios") else { return }
        print("[TerseAuth] Verifying license for: \(userId)")

        URLSession.shared.dataTask(with: url) { [weak self] data, response, error in
            guard let self = self else { return }

            if let error = error {
                print("[TerseAuth] License API error: \(error.localizedDescription)")
                return
            }

            guard let data = data else {
                print("[TerseAuth] No data from license API")
                return
            }

            let rawStr = String(data: data, encoding: .utf8) ?? "nil"
            print("[TerseAuth] License API response: \(rawStr)")

            guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                print("[TerseAuth] Failed to parse JSON")
                return
            }

            DispatchQueue.main.async {
                let serverTier = json["tier"] as? String ?? "free"
                self.status = json["status"] as? String ?? "active"

                // Don't downgrade from Pro if local UserDefaults says user has active subscription
                let localTier = self.defaults.string(forKey: "licenseTier") ?? "free"
                let localOptPerWeek = self.defaults.integer(forKey: "optimizationsPerWeek")
                if localTier == "pro" && localOptPerWeek == -1 && serverTier == "free" {
                    // StoreKit has active subscription — keep Pro, server will catch up
                    self.tier = "pro"
                    self.optimizationsPerWeek = -1
                    self.remaining = -1
                    print("[TerseAuth] Server says free but StoreKit says Pro — keeping Pro")
                } else {
                    self.tier = serverTier
                    if let limits = json["limits"] as? [String: Any] {
                        self.optimizationsPerWeek = limits["optimizations_per_week"] as? Int ?? 120
                        self.maxSessions = limits["max_sessions"] as? Int ?? 1
                    }
                    print("[TerseAuth] Tier: \(self.tier), Status: \(self.status)")
                }

                self.loadWeeklyUsage()
                self.saveLicense()
            }
        }.resume()
    }

    /// Reload usage from UserDefaults (keyboard extension writes here)
    func reloadUsage() {
        defaults.synchronize()
        loadWeeklyUsage()
        objectWillChange.send()
    }

    private func loadWeeklyUsage() {
        let currentWeek = Self.currentWeekString()
        let savedWeek = defaults.string(forKey: "usageWeek") ?? ""
        if savedWeek != currentWeek {
            weeklyUsage = 0
            defaults.set(currentWeek, forKey: "usageWeek")
            defaults.set(0, forKey: "weeklyUsage")
        } else {
            weeklyUsage = defaults.integer(forKey: "weeklyUsage")
        }
        updateRemaining()
    }

    func recordOptimization() {
        let currentWeek = Self.currentWeekString()
        let savedWeek = defaults.string(forKey: "usageWeek") ?? ""
        if savedWeek != currentWeek {
            weeklyUsage = 0
            defaults.set(currentWeek, forKey: "usageWeek")
        }
        weeklyUsage += 1
        defaults.set(weeklyUsage, forKey: "weeklyUsage")
        updateRemaining()
    }

    private func updateRemaining() {
        if optimizationsPerWeek < 0 {
            remaining = -1
        } else {
            remaining = max(0, optimizationsPerWeek - weeklyUsage)
        }
    }

    private func saveLicense() {
        defaults.set(tier, forKey: "licenseTier")
        defaults.set(optimizationsPerWeek, forKey: "optimizationsPerWeek")
        defaults.set(maxSessions, forKey: "maxSessions")
    }

    private func loadLicense() {
        tier = defaults.string(forKey: "licenseTier") ?? "free"
        optimizationsPerWeek = defaults.object(forKey: "optimizationsPerWeek") as? Int ?? 120
        maxSessions = defaults.object(forKey: "maxSessions") as? Int ?? 1
        loadWeeklyUsage()
    }

    private static func currentWeekString() -> String {
        let cal = Calendar(identifier: .iso8601)
        let comps = cal.dateComponents([.yearForWeekOfYear, .weekOfYear], from: Date())
        return "\(comps.yearForWeekOfYear ?? 0)\(String(format: "%02d", comps.weekOfYear ?? 0))"
    }
}

// MARK: - ASWebAuthenticationSession Context Provider

class WebAuthContextProvider: NSObject, ASWebAuthenticationPresentationContextProviding {
    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        return ASPresentationAnchor()
    }
}

File: android-app/app/src/main/java/com/pruneai/terse/auth/TerseAuth.kt
package com.pruneai.terse.auth

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.util.Calendar

class TerseAuth private constructor(context: Context) {

    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val apiBase = "https://www.terseai.org"

    private val _isSignedIn = MutableStateFlow(prefs.getBoolean("isSignedIn", false))
    val isSignedIn: StateFlow<Boolean> = _isSignedIn

    private val _signInError = MutableStateFlow<String?>(null)
    val signInError: StateFlow<String?> = _signInError

    private val _clerkUserId = MutableStateFlow(prefs.getString("clerkUserId", null))
    val clerkUserId: StateFlow<String?> = _clerkUserId

    private val _email = MutableStateFlow(prefs.getString("email", null))
    val email: StateFlow<String?> = _email

    private val _firstName = MutableStateFlow(prefs.getString("firstName", null))
    val firstName: StateFlow<String?> = _firstName

    private val _imageUrl = MutableStateFlow(prefs.getString("imageUrl", null))
    val imageUrl: StateFlow<String?> = _imageUrl

    private val _tier = MutableStateFlow(prefs.getString("licenseTier", "free") ?: "free")
    val tier: StateFlow<String> = _tier

    private val _weeklyUsage = MutableStateFlow(0)
    val weeklyUsage: StateFlow<Int> = _weeklyUsage

    private val _optimizationsPerWeek = MutableStateFlow(prefs.getInt("optimizationsPerWeek", 120))
    val optimizationsPerWeek: StateFlow<Int> = _optimizationsPerWeek

    private val _remaining = MutableStateFlow(0)
    val remaining: StateFlow<Int> = _remaining

    val isUnlimited get() = _optimizationsPerWeek.value < 0
    val canOptimize get() = isUnlimited || _remaining.value > 0
    val tierLabel get() = _tier.value.replaceFirstChar { it.uppercase() }
    val usageText get() = if (isUnlimited) "Unlimited" else "${_remaining.value}/${_optimizationsPerWeek.value} left this week"

    init {
        loadWeeklyUsage()
    }

    fun signInWithGoogle(idToken: String, email: String?, firstName: String?, lastName: String?) {
        scope.launch {
            _signInError.value = null
            try {
                val body = JSONObject().apply {
                    put("identityToken", idToken)
                    put("provider", "google")
                    email?.let { put("email", it) }
                    firstName?.let { put("firstName", it) }
                    lastName?.let { put("lastName", it) }
                }
                val json = post("$apiBase/api/auth/apple", body.toString()) ?: run {
                    _signInError.value = "Network error. Please try again."; return@launch
                }
                if (json.optString("status") != "authenticated") {
                    _signInError.value = "Sign in failed. Please try again."; return@launch
                }
                val userId = json.optString("clerkUserId", "")
                val userEmail = json.optString("email", email ?: "")
                val userImage = json.optString("imageUrl").ifEmpty { null }
                val userFirst = json.optString("firstName").ifEmpty { firstName }
                withContext(Dispatchers.Main) {
                    saveAuth(userId, userEmail, userImage, userFirst)
                    verifyLicense()
                }
            } catch (e: Exception) {
                _signInError.value = "Network error. Please try again."
            }
        }
    }

    fun startWebSignIn(onUrl: (String) -> Unit, onError: (() -> Unit)? = null) {
        scope.launch {
            try {
                val json = post("$apiBase/api/auth/start", "{}")
                val token = json?.optString("token")?.ifEmpty { null }
                if (json == null || token == null) {
                    withContext(Dispatchers.Main) {
                        _signInError.value = "Network error. Please try again."
                        onError?.invoke()
                    }
                    return@launch
                }
                prefs.edit().putString("authToken", token).putBoolean("authCompleted", false).apply()
                val authUrl = "$apiBase/auth-callback.html?token=$token&action=signin"
                withContext(Dispatchers.Main) { onUrl(authUrl) }
                pollForAuth(token)
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    _signInError.value = "Network error. Please try again."
                    onError?.invoke()
                }
            }
        }
    }

    fun checkPendingAuth() {
        if (_isSignedIn.value) return
        val token = prefs.getString("authToken", null) ?: return
        if (prefs.getBoolean("authCompleted", false)) return
        scope.launch { pollForAuth(token) }
    }

    private suspend fun pollForAuth(token: String) {
        delay(1000)
        try {
            val json = get("$apiBase/api/auth/poll/$token") ?: run { pollForAuth(token); return }
            when (json.optString("status")) {
                "authenticated" -> {
                    val userId = json.optString("clerkUserId")
                    val email = json.optString("email")
                    val imageUrl = json.optString("imageUrl").ifEmpty { null }
                    val firstName = json.optString("firstName").ifEmpty { null }
                    prefs.edit().putBoolean("authCompleted", true).apply()
                    withContext(Dispatchers.Main) {
                        saveAuth(userId, email, imageUrl, firstName)
                        verifyLicense()
                    }
                }
                "pending" -> pollForAuth(token)
            }
        } catch (e: Exception) {
            delay(2000); pollForAuth(token)
        }
    }

    fun startCheckout(onUrl: (String) -> Unit) {
        val userId = _clerkUserId.value?.takeIf { it.isNotEmpty() } ?: return
        val userEmail = _email.value ?: ""
        scope.launch {
            try {
                val body = JSONObject().apply {
                    put("tier", "pro")
                    put("clerkUserId", userId)
                    put("clerkUserEmail", userEmail)
                    put("noTrial", false)
                }
                val json = post("$apiBase/api/checkout", body.toString()) ?: return@launch
                val url = json.optString("url").ifEmpty { return@launch }
                withContext(Dispatchers.Main) { onUrl(url) }
            } catch (e: Exception) { /* silent */ }
        }
    }

    fun verifyLicense() {
        val userId = _clerkUserId.value?.takeIf { it.isNotEmpty() } ?: return
        scope.launch {
            try {
                val json = get("$apiBase/api/license/$userId?platform=android") ?: return@launch
                withContext(Dispatchers.Main) {
                    val serverTier = json.optString("tier", "free")
                    val localTier = prefs.getString("licenseTier", "free")
                    val localOpt = prefs.getInt("optimizationsPerWeek", 120)
                    if (localTier == "pro" && localOpt == -1 && serverTier == "free") {
                        _tier.value = "pro"
                        _optimizationsPerWeek.value = -1
                        _remaining.value = -1
                    } else {
                        _tier.value = serverTier
                        val limits = json.optJSONObject("limits")
                        _optimizationsPerWeek.value = limits?.optInt("optimizations_per_week", 120) ?: 120
                    }
                    loadWeeklyUsage()
                    saveLicense()
                }
            } catch (e: Exception) { /* silent */ }
        }
    }

    fun recordOptimization() {
        val currentWeek = currentWeekString
        val savedWeek = prefs.getString("usageWeek", "")
        if (savedWeek != currentWeek) {
            _weeklyUsage.value = 0
            prefs.edit().putString("usageWeek", currentWeek).apply()
        }
        _weeklyUsage.value += 1
        prefs.edit().putInt("weeklyUsage", _weeklyUsage.value).apply()
        updateRemaining()
    }

    fun signOut() {
        _isSignedIn.value = false
        _clerkUserId.value = null
        _email.value = null
        _imageUrl.value = null
        _firstName.value = null
        _tier.value = "free"
        _optimizationsPerWeek.value = 120
        _remaining.value = 120
        _weeklyUsage.value = 0
        prefs.edit()
            .remove("isSignedIn").remove("clerkUserId").remove("email")
            .remove("imageUrl").remove("firstName").remove("licenseTier")
            .remove("optimizationsPerWeek").remove("weeklyUsage").remove("usageWeek")
            .apply()
    }

    fun deleteAccount() {
        val userId = _clerkUserId.value?.takeIf { it.isNotEmpty() }
        if (userId != null) {
            scope.launch {
                try {
                    val body = JSONObject().put("clerkUserId", userId).toString()
                    post("$apiBase/api/auth/delete", body)
                } catch (e: Exception) { /* silent */ }
            }
        }
        signOut()
        prefs.edit()
            .remove("stats_entries").remove("totalTokensSaved")
            .remove("totalTokensOptimized").remove("totalOptimizations")
            .apply()
    }

    private fun saveAuth(clerkUserId: String, email: String, imageUrl: String?, firstName: String?) {
        _isSignedIn.value = true
        _clerkUserId.value = clerkUserId
        _email.value = email
        _imageUrl.value = imageUrl
        _firstName.value = firstName
        prefs.edit()
            .putBoolean("isSignedIn", true)
            .putString("clerkUserId", clerkUserId)
            .putString("email", email)
            .putString("imageUrl", imageUrl)
            .putString("firstName", firstName)
            .apply()
    }

    private fun loadWeeklyUsage() {
        val currentWeek = currentWeekString
        val savedWeek = prefs.getString("usageWeek", "")
        if (savedWeek != currentWeek) {
            _weeklyUsage.value = 0
            prefs.edit().putString("usageWeek", currentWeek).putInt("weeklyUsage", 0).apply()
        } else {
            _weeklyUsage.value = prefs.getInt("weeklyUsage", 0)
        }
        updateRemaining()
    }

    private fun updateRemaining() {
        _remaining.value = if (_optimizationsPerWeek.value < 0) -1
        else maxOf(0, _optimizationsPerWeek.value - _weeklyUsage.value)
    }

    private fun saveLicense() {
        prefs.edit()
            .putString("licenseTier", _tier.value)
            .putInt("optimizationsPerWeek", _optimizationsPerWeek.value)
            .apply()
    }

    private fun post(urlStr: String, body: String): JSONObject? {
        val conn = URL(urlStr).openConnection() as HttpURLConnection
        return try {
            conn.requestMethod = "POST"
            conn.setRequestProperty("Content-Type", "application/json")
            conn.doOutput = true
            conn.connectTimeout = 10000
            conn.readTimeout = 10000
            OutputStreamWriter(conn.outputStream).use { it.write(body) }
            val response = conn.inputStream.bufferedReader().readText()
            JSONObject(response)
        } catch (e: Exception) { null } finally { conn.disconnect() }
    }

    private fun get(urlStr: String): JSONObject? {
        val conn = URL(urlStr).openConnection() as HttpURLConnection
        return try {
            conn.requestMethod = "GET"
            conn.connectTimeout = 10000
            conn.readTimeout = 10000
            val response = conn.inputStream.bufferedReader().readText()
            JSONObject(response)
        } catch (e: Exception) { null } finally { conn.disconnect() }
    }

    companion object {
        const val PREFS_NAME = "terse_auth"
        private val currentWeekString get() = Calendar.getInstance().let {
            "${it.get(Calendar.YEAR)}${it.get(Calendar.WEEK_OF_YEAR).toString().padStart(2, '0')}"
        }

        @Volatile private var INSTANCE: TerseAuth? = null
        fun getInstance(context: Context): TerseAuth =
            INSTANCE ?: synchronized(this) {
                INSTANCE ?: TerseAuth(context.applicationContext).also { INSTANCE = it }
            }
    }
}
